compile.ts

```diff
@@ -603,9 +603,88 @@ class Compiler {
       }
       return dest;
     } else if (ts.isStringLiteral(e)) {
-      return new Variable(new LiteralValue({ id: e.text }));
+      const value = new LiteralValue({ id: e.text });
+      if (dest) {
+        this.#emit(
+          methods.setReg,
+          value,
+          this.ref(dest, VariableOperations.Write)
+        );
+      } else {
+        return new Variable(value);
+      }
+      return dest;
+    } else if (ts.isArrayLiteralExpression(e)) {
+      let arr: number[] = [];
+      for(const property of e.elements) {
+        const value = this.compileExpr(property);
+        if(value.reg?.type !== 'value' || value.reg.value.num == null) {
+          this.#error(`unsupported property ${property.kind}`, property);
+        }
+
+        arr.push(value.reg.value.num);
+      }
+
+      const value = new LiteralValue({
+        coord: {
+          x: arr[0],
+          y: arr[1]
+        }
+      });
+      if (dest) {
+        this.#emit(
+          methods.setReg,
+          value,
+          this.ref(dest, VariableOperations.Write)
+        );
+      } else {
+        return new Variable(value);
+      }
+      return dest;
+    } else if (ts.isObjectLiteralExpression(e)) {
+      let obj = {};
+      for(const property of e.properties) {
+        const name = property.name;
+        if(!name) {
+          this.#error(`property missing name`, property);
+        }
+
+        if (ts.isPropertyAssignment(property)) {
+          const value = this.compileExpr(property.initializer);
+          if(value.reg?.type !== 'value') {
+            this.#error(`unsupported property ${property.kind}`, property);
+          }
+
+          obj = {
+            ...obj,
+            ...value.reg.value
+          }
+        } else {
+          this.#error(`unsupported property ${property.kind}`, property);
+        }
+      }
+
+      const value = new LiteralValue(obj);
+      if (dest) {
+        this.#emit(
+          methods.setReg,
+          value,
+          this.ref(dest, VariableOperations.Write)
+        );
+      } else {
+        return new Variable(value);
+      }
+      return dest;
     } else if (ts.isParenthesizedExpression(e)) {
       return this.compileExpr(e.expression, dest);
+    } else if (ts.isAsExpression(e)) {
+      return this.compileExpr(e.expression, dest);
+    } else if (ts.isPrefixUnaryExpression(e)) {
+      if(e.operator == ts.SyntaxKind.PlusToken) {
+        return this.compileExpr(e.operand, dest);
+      } else {
```

Comment:
Should we support unary `-`? (e.g compile it to a subtraction op)

Reply:
I'd like to, and I have a branch that does this, but I'd like to combine it with literal evaluating math ops before emitting them.

So we could let a user write

```
const v: Value = -1
```

And we could just compile this to `-1` instead of a `sub` operation which "wastes" an instruction (which matters in locked mode, which is the only way to play on the recent servers).

So yes, I'd like to, and I'd like to PR this + my inlining behavior soon. Needs a bunch of cleanup, but a sneak preview: https://github.com/swazrgb/desynced-tools/blob/inlining/compile.ts#L1157-L1178

Might be a good idea to discuss this in a new issue, but do you have any thoughts for the right place to perform such optimizations? During the compilation phase? In a pass afterwards? In the assembler?

Reply:
I think that'd be best in a pass after compilation and before assembly.